#PBS -N Bowtie2
#PBS -l nodes=1:ppn=5
#PBS -l mem=50gb
#PBS -l walltime=12:00:00
#PBS -q inferno     
#PBS -A GT-ktk3-CODA20
#PBS -o 00_log/04f_bowtie2_${n}.out
#PBS -e 00_log/04f_bowtie2_${n}.err

cd $PBS_O_WORKDIR

module purge
module load intel/19.0.5 bowtie2/2.3.5.1

ref="${ref}"

# don't forget to build the bowtie2 index for the reference if it doesn't already exist
#ndx=`echo $ref | cut -d. -f1` # for MAGs step 09
ndx=`echo $ref | cut -d. -f1-4` # for metagenome assemblies step 08
if [ ! -s ${ndx}.rev.1.bt2 ]; then
       bowtie2-build -f $ref $ndx
fi

#ndx=`echo $ref | cut -d. -f1`
#bowtie2 -p 20 -f -x $ndx --interleaved ${qry} -S $out

query="${query}"
out="${out}"

# Run coupled reads as unpaired for abundance mapping should be fine.
bowtie2 -p 5 -f -x $ndx -U ${query} -S $out

echo ref: $ref
echo ndx: $ndx
echo query: $query
echo out: $out

## Run Log ##
## qsub -v ref=MAG.fna,qry=metagenome.fna,out=out.sam,name=sample_name ../00b_PBS/06d_Bowtie2.pbs
## for f in ../01a_Pipeline_Results/02_trim/*.fa; do n=`basename $f | cut -d. -f1`; qsub -v query=$f,ref=01_RepMAGs/01_Concatenated_RepMAGs.fna,out=05_Bowtie2_Mapping/${n}_RepMags.sam,n=$n ../00b_PBS/xxf_Bowtie2.pbs; done

##### WAS THROWING SEGV ERROR with interlevaed read issue.
##### SAME ERROR WITH -U unparied reads
##### added module purge first and -U worked
##### Did not try interleaved again yet.
